### ISSUE TITLE
pgwire: multiple active result sets (portals) not supported

### ISSUE THREAD

**Issue by jordanlewis** [cockroachdb/cockroach#40195]
CockroachDB does not currently support multiple active pgwire portals on the same session. This means that you can't perform the following sequence of events:

1. open portal for query a, which will eventually return 1000 rows
2. ask for 500 rows of query a
3. open portal for query b
4. ask for 500 rows of query b
5. ask for 500 rows of query a

This sequence will fail at step 3, unless the user closes the portal for query a first, in which case this sequence will fail at step 5.

**Comment by rafiss** [comment]
The pgjcbc test `org.postgresql.test.jdbc2.CursorFetchTest.testGetRow` used to fail because of this lack of support, but that test started passing. I don't know why. #41366

**Comment by imalsogreg** [comment]
We (myself and @robo-corg) hit an edge-case related to this issue. When we use the rust [sqlx](https://github.com/launchbadge/sqlx) postgres bindings, queries run inside of transactions trigger the "multiple portals" issue. This happens even when we run just a single query, which we think creates just one portal.

![wireshark_sqlx](https://user-images.githubusercontent.com/993484/103315304-316c6500-49f3-11eb-8ba7-63d2bf63ed7d.png)

Here's a wireshark/tcpdump of a session with sqlx that exhibits this behavior. We can attach the .pcap file if it's helpful.

**Comment by rafiss** [comment]
@imalsogreg Thanks for sharing this case! A wireshark .pcap would definitely be very helpful for us. We can use that to more easily write a test within our project.

**Comment by imalsogreg** [comment]
@rafiss Cool. Here's the .pcap from when I ran the minimal reproducing case. It corresponds to the left panel in the screenshot above.

[packets4.pcap.zip](https://github.com/cockroachdb/cockroach/files/5754348/packets4.pcap.zip)

**Comment by bka9** [comment]
After upgrading from 20.1.12 to 20.2.7 our use of snaplogic to query data has failed with the following stack trace:
```
com.snaplogic.snap.api.SnapDataException: Failed to execute select query.
	at com.snaplogic.snaps.sql.SimpleSqlSnap.process(SimpleSqlSnap.java:399)
	at com.snaplogic.snap.api.ExecutionUtil.process(ExecutionUtil.java:106)
	at com.snaplogic.snap.api.ExecutionUtil.executeForDocument(ExecutionUtil.java:118)
	at com.snaplogic.snap.api.ExecutionUtil.execute(ExecutionUtil.java:81)
	at com.snaplogic.snap.api.SimpleSnap.execute(SimpleSnap.java:70)
	at com.snaplogic.cc.snap.common.SnapRunnableImpl.executeSnap(SnapRunnableImpl.java:770)
	at com.snaplogic.cc.snap.common.SnapRunnableImpl.execute(SnapRunnableImpl.java:552)
	at com.snaplogic.cc.snap.common.SnapRunnableImpl.doRun(SnapRunnableImpl.java:836)
	at com.snaplogic.cc.snap.common.SnapRunnableImpl.call(SnapRunnableImpl.java:402)
	at com.snaplogic.cc.snap.common.SnapRunnableImpl.call(SnapRunnableImpl.java:116)
	at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
	at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
Caused by: com.snaplogic.api.ExecutionException: Failed to execute select query.
	at com.snaplogic.snap.api.sql.operations.JdbcOperationsImpl.select(JdbcOperationsImpl.java:700)
	at com.snaplogic.snaps.sql.SimpleSqlSelectSnap.selectAndSpoolResults(SimpleSqlSelectSnap.java:714)
	at com.snaplogic.snaps.sql.SimpleSqlSelectSnap.lambda$select$5(SimpleSqlSelectSnap.java:699)
	at net.jodah.failsafe.Functions$12.call(Functions.java:274)
	at net.jodah.failsafe.SyncFailsafe.call(SyncFailsafe.java:145)
	at net.jodah.failsafe.SyncFailsafe.run(SyncFailsafe.java:93)
	at com.snaplogic.snaps.sql.SimpleSqlSelectSnap.select(SimpleSqlSelectSnap.java:696)
	at com.snaplogic.snaps.sql.SimpleSqlSelectSnap.processDocument(SimpleSqlSelectSnap.java:678)
	at com.snaplogic.snaps.sql.SimpleSqlSnap.process(SimpleSqlSnap.java:390)
	... 15 more
	Suppressed: org.jooq.exception.DataAccessException: SQL [null]; ERROR: current transaction is aborted, commands ignored until end of transaction block
		at org.jooq_3.9.1.POSTGRES.debug(Unknown Source)
		at org.jooq.impl.Tools.translate(Tools.java:1983)
		at org.jooq.impl.MetaDataFieldProvider.init(MetaDataFieldProvider.java:158)
		at org.jooq.impl.MetaDataFieldProvider.<init>(MetaDataFieldProvider.java:83)
		at org.jooq.impl.SelectQueryImpl.getFields(SelectQueryImpl.java:407)
		at org.jooq.impl.AbstractResultQuery.execute(AbstractResultQuery.java:285)
		at org.jooq.impl.AbstractQuery.execute(AbstractQuery.java:349)
		at org.jooq.impl.AbstractResultQuery.fetchLazy(AbstractResultQuery.java:368)
		at com.snaplogic.snap.api.sql.operations.JdbcOperationsImpl.fetchLazy(JdbcOperationsImpl.java:841)
		at com.snaplogic.snap.api.sql.operations.JdbcOperationsImpl.select(JdbcOperationsImpl.java:694)
		... 23 more
	Caused by: org.postgresql.util.PSQLException: ERROR: current transaction is aborted, commands ignored until end of transaction block
		at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2468)
		at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2211)
		at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:309)
		at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:446)
		at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:370)
		at org.postgresql.jdbc.PgStatement.executeWithFlags(PgStatement.java:311)
		at org.postgresql.jdbc.PgStatement.executeCachedSql(PgStatement.java:297)
		at org.postgresql.jdbc.PgStatement.executeWithFlags(PgStatement.java:274)
		at org.postgresql.jdbc.PgStatement.executeQuery(PgStatement.java:225)
		at org.postgresql.jdbc.PgResultSetMetaData.fetchFieldMetaData(PgResultSetMetaData.java:245)
		at org.postgresql.jdbc.PgResultSetMetaData.isAutoIncrement(PgResultSetMetaData.java:57)
		at org.postgresql.jdbc.PgResultSetMetaData.getColumnTypeName(PgResultSetMetaData.java:318)
		at org.jooq.impl.MetaDataFieldProvider.init(MetaDataFieldProvider.java:129)
		... 30 more
	Caused by: org.postgresql.util.PSQLException: ERROR: unimplemented: multiple active portals not supported
  Hint: You have attempted to use a feature that is not yet implemented.
See: https://go.crdb.dev/issue-v/40195/v20.2
		at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2468)
		at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2211)
		at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:309)
		at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:446)
		at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:370)
		at org.postgresql.jdbc.PgStatement.executeWithFlags(PgStatement.java:311)
		at org.postgresql.jdbc.PgStatement.executeCachedSql(PgStatement.java:297)
		at org.postgresql.jdbc.PgStatement.executeWithFlags(PgStatement.java:274)
		at org.postgresql.jdbc.PgStatement.executeQuery(PgStatement.java:225)
		at org.postgresql.jdbc.PgResultSetMetaData.fetchFieldMetaData(PgResultSetMetaData.java:245)
		at org.postgresql.jdbc.PgResultSetMetaData.getBaseTableName(PgResultSetMetaData.java:289)
		at org.postgresql.jdbc.PgResultSetMetaData.getTableName(PgResultSetMetaData.java:285)
		at org.jooq.impl.MetaDataFieldProvider.init(MetaDataFieldProvider.java:112)
		... 30 more
Reason: SQL [null]; ERROR: current transaction is aborted, commands ignored until end of transaction block
Resolution: Please check sql query.
```

The query is a simple select statement.

**Comment by rafiss** [comment]
I'm coming back to this now. It seems like the error message is incorrectly used if you execute a statement **with a row count limit** within a transaction. Here's a test case that can be put in `pkg/sql/pgwire/testdata/pgtest/` that reproduces.

```
send
Query {"String": "BEGIN"}
----

until
ReadyForQuery
----
{"Type":"CommandComplete","CommandTag":"BEGIN"}
{"Type":"ReadyForQuery","TxStatus":"T"}

# 83 = ASCII 'S' for Statement
# 50 = ASCII '2'
# ParameterFormatCodes = [0] for text format
send
Parse {"Name": "s6", "Query": "select $1::int8", "ParameterOIDs": [20]}
Describe {"ObjectType": 83, "Name": "s6"}
Bind {"DestinationPortal": "p6", "PreparedStatement": "s6", "ParameterFormatCodes": [0], "ResultFormatCodes": [0], "Parameters": [[50]]}
Execute {"Portal": "p6", "MaxRows": 1}
Sync
----

until
ReadyForQuery
----
{"Type":"ParseComplete"}
{"Type":"ParameterDescription","ParameterOIDs":[20]}
{"Type":"RowDescription","Fields":[{"Name":"int8","TableOID":0,"TableAttributeNumber":0,"DataTypeOID":20,"DataTypeSize":8,"TypeModifier":-1,"Format":0}]}
{"Type":"BindComplete"}
{"Type":"DataRow","Values":[{"text":"2"}]}
{"Type":"PortalSuspended"}
{"Type":"ReadyForQuery","TxStatus":"T"}

send
Query {"String": "COMMIT"}
----

until
ReadyForQuery
----
{"Type":"CommandComplete","CommandTag":"COMMIT"}
{"Type":"ReadyForQuery","TxStatus":"I"}
```

It works against Postgres using `make test PKG=./pkg/sql/pgwire TESTS=TestPGTest TESTFLAGS="-addr=localhost:5432 -user=rafiss"`

But it fails against CockroachDB `make test PKG=./pkg/sql/pgwire TESTS=TestPGTest`

```
got &pgproto3.ErrorResponse{Severity:"ERROR", SeverityUnlocalized:"", Code:"0A000", Message:"unimplemented: multiple active portals not supported", Detail:"", Hint:"You have attempted to use a feature that is not yet implemented.\nSee: https://go.crdb.dev/issue-v/40195/v21.1", Position:0, InternalPosition:0, InternalQuery:"", Where:"", SchemaName:"", TableName:"", ColumnName:"", DataTypeName:"", ConstraintName:"", File:"distsql_running.go", Line:775, Routine:"init", UnknownFields:map[uint8]string(nil)}
```

**Comment by rafiss** [comment]
I see that this limitation is described here: https://github.com/cockroachdb/cockroach/issues/42912#issuecomment-566250989 -- I'll see if we can address this now

**Comment by vitruvvius** [comment]
+ 1

**Comment by somin** [comment]
I would love to use cockroach DB however I am also running into this issue in our trial. There are a few workarounds in Rust however their usability is extremely limited and cumbersome to work with. Any potential fixes for this would be much appreciated!

**Comment by IMFIL** [comment]
+1

**Comment by bragaigor** [comment]
+1

**Comment by franic220** [comment]
+1

**Comment by vitruvvius** [comment]
this happens in particular for me when appending

RETURNING *

to an insert query in rust

**Comment by knz** [comment]
Here's an absolutely amazing analysis by @rafiss that explains how kafka-connect runs into this issue (copy-pasting from the private repo):

## Cause
CockroachDB has had a long-standing limitation in its Postgres wire protocol implementation, which predates v20.1. The limitation is that "multiple active portals are not supported." The less technical explanation of the limitation is that a client cannot re-use a connection to initiate a new query until all of the rows from any previous query have been fully read.

#### Here's what is happening in kafka-connect:
1. kafka-connect begins a transaction.
2. kafka-connect creates a prepared statement for the `SELECT ... FROM risk_profile_outbox` query shared previously.
3. kafka-connect sets the JDBC `fetchSize` to `batch.max.rows` (defaults to 100).
4. kafka-connect executes the prepared statement and gets back a ResultSet.
5. Before reading the results of the prepared statement, kafka-connect calls its `describeColumn` method, which runs the pgjdbc `fetchFieldMetadata` method which causes a metadata query to be sent to the database.

#### Here's what the problem is:
- When step (5) occurs, the portal opened by the prepared statement has not yet been closed.
- Normally, a portal would be closed by calling `resultSet.close()`, but that can't happen here because the results haven't been read yet.
- CockroachDB is able to automatically close the portal even without the client using `resultSet.close()` in some cases: specifically, CockroachDB can close the portal only if the `fetchSize` is **strictly larger** than the number of rows returned by the query.
- Since the portal is still open when another query is sent in step (5), the application hits the "multiple active portals not supported" error.

####  So the question is why does the error happen with CRDB v20.2 and not CRDB v20.1?
- It is because v20.2 includes this change: https://github.com/cockroachdb/cockroach/pull/51479. The summary of that change is that CRDB will report the `tableOID` when returning data to the client inside of a transaction. Previously it did not do so, which was a divergence from the Postgres wire protocol.
- The change above affects the PGJDBC driver in an important way: it means that the PGJDBC will start running the `fetchFieldMetaData` method (on v20.2), whereas in v20.1 it would short-circuit and not run the metadata query.
    - See this line of PGJDBC https://github.com/pgjdbc/pgjdbc/blob/efa57b3a56d277284ef5f7d4417057d1f86ae0e8/pgjdbc/src/main/java/org/postgresql/core/Field.java#L81 -- it shows that when no table OID is reported, the FieldMetadata is initialized to a default value. When a table OID **is** reported, the FieldMetadata is initialized to null.
    - This affects the `fetchFieldMetadata` as follows: https://github.com/pgjdbc/pgjdbc/blob/efa57b3a56d277284ef5f7d4417057d1f86ae0e8/pgjdbc/src/main/java/org/postgresql/jdbc/PgResultSetMetaData.java#L192-L194 -- when the FieldMetadata is null, the method does not short-circuit so it runs the metadata query.
    - As described above, executing the metadata query triggers the error.
- Unfortunately, the CockroachDB change [#51479](https://github.com/cockroachdb/cockroach/pull/51479) does not appear in **any** release note, so we can't point the customer to any docs on this.
    - Though do note that the reason this PR was originally created was because a different Java tool was broken due to CRDB's lack of reporting the `tableOID` correctly.

#### Hang on, does this mean anyone who uses Java is going to see this error?
No. Although the PGJDBC _can_ try to use multiple active portals, it is not the default behavior. Specifically, all of the following must be true of the tool/application:
- Uses a prepared statement.
- Sets the `fetchSize` of the statement to a positive number.
- After running the statement, the database returns more rows than the `fetchSize`.
- The client runs some other query **before** calling `close()` on  the ResultSet.

It so happens that kafka-connect does all of those things. We have not done any testing of kafka-connect internally.

## Immediate workaround
- Set the kafka-connect setting `batch.max.rows` to a value larger than the number of rows returned by the query. It must be strictly larger than the number of rows in order for CockroachDB to be able to know that the portal can be closed automatically. 
    - The `batch.max.rows` setting cannot be **equal** to the number of rows. The portal will still remain open in that case.
    - Also, I previously advised setting the `batch.max.rows` setting to 0. Although this does disable the `fetchSize` in PGJDBC, it doesn't work because kafka-connect does not handle a value of 0. See the following code, which shows that kafka-connect will fail to read any results if `batch.max.rows` is 0: https://github.com/confluentinc/kafka-connect-jdbc/blob/f3414dd9f0a3223b0142de4a74bf85500c1b0ea4/src/main/java/io/confluent/connect/jdbc/source/JdbcSourceTask.java#L373-L377
    - If the query they are using has a LIMIT, then set `batch.max.rows` to one more than the LIMIT. If the query does not have a LIMIT, then set `batch.max.rows` to a large value that is assured to be larger than the number of rows returned by the query.

## Longer term fix
Get https://github.com/cockroachdb/cockroach/issues/40195 onto the Product roadmap.

**Comment by rytaft** [comment]
cc @vy-ton for prioritization

**Comment by dilip2048** [comment]
Is this fixed?

**Comment by klesniewski** [comment]
We had a case, where we were receiving this error. The cause however had nothing to do with multiple portals. The problem was occurring also with v21.1.11.

We have a query that fetches data (around 50-60k rows) in batches of 10k using `fetchSize`. Every now and then the query was failing with:
```
ERROR: unimplemented: multiple active portals not supported
org.postgresql.util.PSQLException: ERROR: unimplemented: multiple active portals not supported
```
We have analyzed the code but didn't see how we could have multiple portals open:

1. We take a connection from connection pool.
2. We open a transaction.
3. We make a select statement which fetches data with `fetchSize` of 10k.
4. We iterate over the results and process them.
5. We correctly cleanup cursor even in case of exception.
6. We close the transaction (even in case of exception).
7. We return the connection to connection pool.

There is no other query we do in the transaction.

Now, we found the problem a bit by accident. Turns out that the query, if running long enough, can fail with [ReadWithUncertaintyIntervalError](https://www.cockroachlabs.com/docs/v21.1/transaction-retry-error-reference.html#readwithinuncertaintyinterval):

![image](https://user-images.githubusercontent.com/6713429/140145136-6b5e999c-5b00-4e35-b80b-962edc2ee483.png)

The select we do spans a table, which is updated multiple times per second. Our best bet was that this is what is causing the multiple active portals error. We have changed the transaction, so that it runs `AS OF SYSTEM TIME follower_read_timestamp()`. We have this change in production for 3 days and we haven't observed this error anymore, whereas previously we had couple of instances per day.

**Comment by rafiss** [comment]
Thanks for the write up. In that case our logic to automatically close the portal when a transaction finishes did not kick in, since there was an error when trying to commit the transaction.

**Comment by klesniewski** [comment]
We were getting the error not when committing the transaction, but when fetching next portion of rows, if that helps.

To be honest, my knowledge here is limited, so I hope the extra details will help you see if it is a problem with multiple portals or whether this error can be thrown also due to contention.

When we got the error, we first tried to understand what a portal is and what multiple active portals mean. With what we read in PostgreSQL documentation, we tried to find where multiple portals could be created, but everything seemed fine. The query (within explicit transaction block) goes like this (implemented in Kotlin with jOOQ):

```kotlin
.orderBy(assetTable.ID)
.fetchSize(jdbcFetchSize) // prevent loading the whole result set in memory
.fetchLazy().use {
     while (it.hasNext()) {
        processSearchRecord(it.fetchNext()!!, resultMap, geoHashPrecision)
      }
}
```

This should correctly take care of closing the cursor (and portal?) before we even reach the end of transaction.

**Comment by rathboma** [comment]
I'm also having this issue, but with a single result set in nodejs.

This is the code: It's when reading from a cursor. Here is the code that is failing:

https://github.com/beekeeper-studio/beekeeper-studio/blob/master/apps/studio/src/lib/db/clients/postgresql/PsqlCursor.ts#L48

Is this pattern possible with cockroach? Like can I change this to make it work, or is this a lost cause?

**Comment by vitruvvius** [comment]
the underlying library typically can avoid this - in my case, sqlx in rust. you had to call fetch_all vs fetch_optional / fetch_one. not sure if it has changed since i continue always w/ fetch_all.

it's worth it if you're looking for strict serializability at scale.

**Comment by rafiss** [comment]
@rathboma the other reason this might be happening is if you previously were reading from a different cursor, but did not read all the results until the cursor was completed. Is it possible your code might be doing that?

If your code can't read all the results until the cursor is exhausted, then another solution is to make sure you *always* call `cursor.Close` (https://node-postgres.com/api/cursor#close) when you are done reading from a cursor.

**Comment by rathboma** [comment]
@rafiss It's possible, but unlikely, as I can replicate the problem when running a single unit test. That runs a bunch of one-off set-up queries, then tries to select with a single cursor.

**Comment by rafiss** [comment]
Ah I see, if you can provide instructions on how I can run these tests, I can take a Wireshark network capture to see exactly what's happening at the pgwire level. (Or if you're able to provide a Wireshark capture while running, that'd help too.)

**Comment by rathboma** [comment]
I've never used wireshark, but can give it a go.

Thanks so much for helping to debug. Here are the instructions. I've created a branch which will only run that single test:

```bash
git clone git@github.com:beekeeper-studio/beekeeper-studio.git
cd beekeeper-studio
git checkout cockroach-stream-fix
yarn install
yarn workspace beekeeper-studio test:integration apps/studio/tests/integration/lib/db/clients/cockroach.spec.js
```
It requires:
- docker to be installed
- ability to build native sqlite binary (x86.64 or arm64)

**Comment by rafiss** [comment]
@rathboma  Thanks for those instructions. I am getting a different error. (Also I had to switch to node14 since it wouldn't work with node16). To keep this issue more focused, let's move the conversation here: #74187

**Comment by mcanalesmayo** [comment]
Are there any plans to fix this? We are running into this issue as well with a very simple scenario where we are reading and updating rows inside a transaction, using Vertx streaming API (batches).

**Comment by mari-crl** [comment]
:information_source: Hello! I am a human and not at all a robot! Look at my very human username! :robot: :notes:
:thinking: Although I tried very hard to figure out what to do with this issue, more powerful human brains will need to help me. (specifically: Both Github and Jira issues were modified after exalate problems) :confounded:
:arrows_counterclockwise: **Please visit this issue's mirror at [CRDB-5559](https://cockroachlabs.atlassian.net/browse/CRDB-5559) and try to sync the two sides up manually.** :star2:
:white_check_mark: When you're finished, comment saying as much asn a member of Developer Infrastructure will be along to finish linking. :link:
:no_entry_sign: Note that until this is done, **this issue is not and will not be synced to Jira with Exalate**. :no_entry_sign:
:sweat_smile: Feeling lost? Don't worry about it! A member of @cockroachdb/exalate-22-cleanup-team will be along shortly to help! :+1:
:construction_worker: Developer Infrastructure members: when ready, open Exalate from the right-hand menu of the mirror issue in Jira, then choose Connect and enter this issue's URN: `cockroachdb/cockroach-40195`. Either way, delete this comment when you're done. :key:
:pray: Thank you for your compliance, my fellow humans! :robot: :wave:

**Comment by markddrake** [comment]
I am getting this error attempting to process a table with 107 rows in it using node pg and node pg QueryStream.  As far as I am aware I have no other active queries..

https://github.com/brianc/node-postgres/issues/2767

**Comment by rafiss** [comment]
Thanks for linking that issue. It looks like node-pg is using the Flush message. We should be able to handle that in a portal, even without adding full support for multiple active portals.

**Comment by markddrake** [comment]
If you vids are going to fix this do you want me to open a new issue on cockroach with the  tetscas

**Comment by wb14123** [comment]
I got this error when read a table with 1000 rows per batch and update them in the same transaction. But if I read the table with 100 rows per time, there is no error. This is kind of a strange behavior and is very confusing.

**Comment by lukaseder** [comment]
@bka9 regarding the jOOQ specific issue from your comment here: https://github.com/cockroachdb/cockroach/issues/40195#issuecomment-818393117

It's likely you ran into this because of using a plain SQL query without explicitly specifying the projection, given this line in your stack trace:

> at org.jooq.impl.MetaDataFieldProvider.init(MetaDataFieldProvider.java:112)

It's always good to provide explicit projections (`SELECT` clause) rather than running `SELECT *` queries for other performance reasons. This issue shows that fetching column meta data in a separate query causes additional performance problems. A more detailed description of the jOOQ related problem can be seen here: https://github.com/jOOQ/jOOQ/issues/14392

**Comment by rafiss** [comment]
> I'm building a serverless application, where a lot of threads can run in parallel, but through different instances.
Can it be accomplished with current CockroachDB version?

Based on your description, it sounds like each thread is using a different SQL connection. If that's the case, then yes, this should work in existing CockroachDB versions. 

---

Also an update: in v23.1 CockroachDB will support multiple active portals as a [preview feature](https://www.cockroachlabs.com/docs/stable/cockroachdb-feature-availability.html). The feature will be off by default, and can be enabled by setting the session variable `multiple_active_portals_enabled` to `true`. This will be available in `v23.1.0-beta.2`.

We encourage everyone to try it out for their use cases, with the caveat that there are still known issues and incomplete performance testing. The list of known issues can be found with the [A-pausable-portals label](https://github.com/cockroachdb/cockroach/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc++label%3AA-pausable-portals+), and we always welcome new bug reports.
